lib/javascript/JavascriptParser.js

```diff
@@ -417,23 +417,27 @@ class JavascriptParser extends Parser {
 
 				const left = this.evaluateExpression(expr.left);
 				if (!left) return;
+				let keepRight;
 				if (expr.operator === "&&") {
 					const leftAsBool = left.asBool();
 					if (leftAsBool === false) return left.setRange(expr.range);
-					if (leftAsBool !== true) return;
+					keepRight = leftAsBool === true;
 				} else if (expr.operator === "||") {
 					const leftAsBool = left.asBool();
 					if (leftAsBool === true) return left.setRange(expr.range);
-					if (leftAsBool !== false) return;
+					keepRight = leftAsBool === false;
 				} else if (expr.operator === "??") {
 					const leftAsNullish = left.asNullish();
 					if (leftAsNullish === false) return left.setRange(expr.range);
-					if (leftAsNullish !== true) return;
+					keepRight = leftAsNullish === true;
 				} else return;
 				const right = this.evaluateExpression(expr.right);
 				if (!right) return;
 				if (left.couldHaveSideEffects()) right.setSideEffects();
-				return right.setRange(expr.range);
+
+				if (typeof right.asBool() === "boolean" || keepRight) {
+					return right.setRange(expr.range);
```

Comment:
You can't use the right side if the left side is unknown

e. g. `f() && "string"` would evaluate to `"string"`, but that's not correct.
You must construct a new BasicExpression which is only truthy in that case.